src/c_api/c_api.cc

```diff
@@ -403,6 +406,237 @@ int MXFuncInvokeEx(FunctionHandle fun,
   API_END();
 }
 
+int MXImperativeInvoke(AtomicSymbolCreator creator,
+                       int num_inputs,
+                       NDArrayHandle *inputs,
+                       int *num_outputs,
+                       NDArrayHandle **outputs,
+                       int num_params,
+                       char **param_keys,
+                       char **param_vals) {
+  static auto& num_args = nnvm::Op::GetAttr<std::string>("key_var_num_args");
+  static auto& infershape = nnvm::Op::GetAttr<nnvm::FInferShape>("FInferShape");
+  static auto& infertype = nnvm::Op::GetAttr<nnvm::FInferType>("FInferType");
+  static auto& fcpu = nnvm::Op::GetAttr<FCompute>("FCompute<cpu>");
+  static auto& fgpu = nnvm::Op::GetAttr<FCompute>("FCompute<gpu>");
+  static auto& ndfunc = nnvm::Op::GetAttr<FNDArrayFunction>("FNDArrayFunction");
+  static auto& createop = nnvm::Op::GetAttr<FCreateLayerOp>("FCreateLayerOp");
+  static auto& mutate = nnvm::Op::GetAttr<nnvm::FMutateInputs>("FMutateInputs");
+  static auto& tmp_resource = nnvm::Op::GetAttr<FResourceRequest>("FResourceRequest");
+  const nnvm::Op* op = static_cast<nnvm::Op*>(creator);
+  MXAPIThreadLocalEntry *ret = MXAPIThreadLocalStore::Get();
+
+  API_BEGIN();
+  nnvm::NodeAttrs attrs;
+  for (int i = 0; i < num_params; ++i) {
```

Comment:
any idea to speedup these operations? One way is to put most of vectors in threadlocal workspace to avoid re-allocation